ts/test/session/unit/utils/job_runner/user_sync_job/UserSyncJob_test.ts

```diff
@@ -0,0 +1,408 @@
+    const networkTimestamp = 4444;
+    Sinon.stub(NetworkTime, 'now').returns(networkTimestamp);
+
+    pendingChangesForUsStub.resolves({
+      messages: [profile, contact],
+      allOldHashes: new Set('123'),
+    });
+    const result = await UserSync.pushChangesToUserSwarmIfNeeded();
+
+    sendStub.resolves(undefined);
+    expect(result).to.be.eq(RunJobResult.RetryJobIfPossible); // not returning anything in the sendstub so network issue happened
+    expect(sendStub.callCount).to.be.eq(1);
+    expect(pendingChangesForUsStub.callCount).to.be.eq(1);
+    expect(dump.callCount).to.be.eq(1);
+    expect(dump.firstCall.args).to.be.deep.eq(['ContactsConfig']);
+  });
+
+  it('calls sendEncryptedDataToSnode with the right data x3 and retry if network returned nothing then success', async () => {
+    const profile = userChange(sodium, SnodeNamespaces.UserProfile, 321);
+    const contact = userChange(sodium, SnodeNamespaces.UserContacts, 123);
+    const groups = userChange(sodium, SnodeNamespaces.UserGroups, 111);
+
+    pendingChangesForUsStub.resolves({
+      messages: [profile, contact, groups],
+      allOldHashes: new Set('123'),
+    });
+    const changes: Array<UserSuccessfulChange> = [
+      {
+        pushed: profile,
+        updatedHash: 'hashprofile',
+      },
+      {
+        pushed: contact,
+        updatedHash: 'hashcontact',
+      },
+      {
+        pushed: groups,
+        updatedHash: 'hashgroup',
+      },
+    ];
+    Sinon.stub(LibSessionUtil, 'batchResultsToUserSuccessfulChange').returns(changes);
+    const confirmPushed = Sinon.stub(UserGenericWrapperActions, 'confirmPushed').resolves();
+
+    // all 4 need to be dumped
+    const needsDump = Sinon.stub(UserGenericWrapperActions, 'needsDump').resolves(true);
+
+    // ============ 1st try, let's say we didn't get as much entries in the result as expected. This should be a fail
+    sendStub.resolves([
+      { code: 200, body: { hash: 'hashprofile' } },
+      { code: 200, body: { hash: 'hashcontact' } },
+      { code: 200, body: { hash: 'hashgroup' } },
+    ]);
+    let result = await UserSync.pushChangesToUserSwarmIfNeeded();
+
+    expect(sendStub.callCount).to.be.eq(1);
+    expect(pendingChangesForUsStub.callCount).to.be.eq(1);
+    expect(dump.getCalls().map(m => m.args)).to.be.deep.eq([
+      ['UserConfig'],
+      ['ContactsConfig'],
+      ['UserGroupsConfig'],
+      ['ConvoInfoVolatileConfig'],
+    ]);
+    expect(dump.callCount).to.be.eq(4);
+
+    expect(needsDump.getCalls().map(m => m.args)).to.be.deep.eq([
+      ['UserConfig'],
+      ['ContactsConfig'],
+      ['UserGroupsConfig'],
+      ['ConvoInfoVolatileConfig'],
+    ]);
+    expect(needsDump.callCount).to.be.eq(4);
+
+    expect(confirmPushed.callCount).to.be.eq(0); // first send failed, shouldn't confirm pushed
+    expect(result).to.be.eq(RunJobResult.RetryJobIfPossible);
+
+    // ============= second try: we now should get a success
```

Comment:
Brain is struggling a bit but where is the third attempt?

Reply:
@Bilb missed

Reply:
only 2 of them. I've fixed the title of the text